common/config/config_document_generator.go

```diff
@@ -0,0 +1,543 @@
+func parseStructGodocs(filePath string, lineNumber int) (map[string]string, error) {
+
+	fields := make(map[string]string)
+
+	// Read the file.
+	fileBytes, err := os.ReadFile(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file %q: %w", filePath, err)
+	}
+	fileString := string(fileBytes)
+
+	lines := strings.Split(fileString, "\n")
+	if lineNumber < 1 || lineNumber > len(lines) {
+		return nil, fmt.Errorf("line number %d out of range for file %q with %d lines",
+			lineNumber, filePath, len(lines))
+	}
+
+	var godoc strings.Builder
+
+	// Search for fields starting from the given line number (which should be the line where the struct is defined).
+	for i := lineNumber - 1; i < len(lines); i++ {
+		line := strings.TrimSpace(lines[i])
+		if line == "" {
+			// Skip blank lines, but reset the GoDoc accumulator. We should assume blank lines mean that the prior
+			// GoDoc comments are not associated with the next field.
+			godoc.Reset()
+			continue
+		}
+
+		if strings.Contains(line, "}") {
+			// Anonymous (i.e. nested) structs are prohibited, so we can assume this is the end of the struct.
+			break
+		}
+
+		if strings.HasPrefix(line, "//") {
+			// Accumulate GoDoc comments for the next field.
+			if godoc.Len() > 0 {
+				godoc.WriteString("\n")
+			}
+			godoc.WriteString(strings.TrimSpace(strings.TrimPrefix(line, "//")))
+			continue
+		}
+
+		// We've found a line that isn't a comment or blank line, so it should be a struct field.
+		// Extract the field name and the accumulated GoDoc comments.
+
+		godocString := godoc.String()
+		godoc.Reset()
+
+		parts := strings.Split(line, " ")
+		if len(parts) < 2 {
+			return nil, fmt.Errorf("failed to parse struct field from line %q in file %q", line, filePath)
+		}
+
+		fieldName := strings.TrimSpace(parts[0])
+		if fieldName == "" {
+			return nil, fmt.Errorf("failed to parse struct field from line %q in file %q", line, filePath)
+		}
+
+		fields[fieldName] = godocString
+	}
+
+	return fields, nil
+}
+
+// All the data needed to document a config field.
+type configFieldData struct {
+	// Name of the environment variable that will set this field.
+	EnvVar string
+	// The toml tag that will set this field.
+	TOML string
+	// Type of the field as a string.
+	FieldType string
+	// The default value of the field as a string.
+	DefaultValue string
+	// GoDoc comment associated with the field.
+	Godoc string
+
+	// If true, this field is required.
+	Required bool
+	// If true, this field is deprecated.
+	Deprecated bool
+	// If true, this field is unsafe.
+	Unsafe bool
+}
+
+// parseDocsTag parses the `docs` struct tag and returns whether the field is required, deprecated, or unsafe.
+// Only one tag value is allowed per field.
+func parseDocsTag(tag string) (required bool, deprecated bool, unsafe bool) {
+	if tag == "" {
+		return false, false, false
+	}
+
+	// Trim whitespace for flexibility
+	tag = strings.TrimSpace(tag)
+
+	switch tag {
```

Comment:
Would it make sense to throw an error if an unexpected tag is encountered? Seems such a case would most likely indicate a typo in the tag